test-data/unit/check-literal.test

```diff
@@ -0,0 +1,505 @@
+
+b: Literal["Foo"]
+reveal_type(b)      # E: Revealed type is 'Literal['Foo']'
+
+c: "Literal[Foo]"
+reveal_type(c)      # E: Revealed type is 'Literal[5]'
+
+d: "Literal['Foo']"
+reveal_type(d)      # E: Revealed type is 'Literal['Foo']'
+
+e: Literal[Foo, 'Foo']
+reveal_type(e)      # E: Revealed type is 'Union[Literal[5], Literal['Foo']]'
+
+Foo = Literal[5]
+[out]
+
+[case testLiteralBiasTowardsAssumingForwardReferencesForTypeComments]
+from typing_extensions import Literal
+
+a = None  # type: Foo
+reveal_type(a)      # E: Revealed type is '__main__.Foo'
+
+b = None  # type: "Foo"
+reveal_type(b)      # E: Revealed type is '__main__.Foo'
+
+c = None  # type: Literal["Foo"]
+reveal_type(c)      # E: Revealed type is 'Literal['Foo']'
+
+d = None  # type: Literal[Foo]   # E: Parameter 1 of Literal[...] is invalid
+
+class Foo: pass
+[out]
+
+
+--
+-- Check how we handle very basic subtyping and other useful things
+--
+
+[case testLiteralCallingFunction]
+from typing_extensions import Literal
+def foo(x: Literal[3]) -> None: pass
+
+a: Literal[1]
+b: Literal[2]
+c: int
+
+foo(a)  # E: Argument 1 to "foo" has incompatible type "Literal[1]"; expected "Literal[3]"
+foo(b)  # E: Argument 1 to "foo" has incompatible type "Literal[2]"; expected "Literal[3]"
+foo(c)  # E: Argument 1 to "foo" has incompatible type "int"; expected "Literal[3]"
+[out]
+
+[case testLiteralCallingFunctionWithUnionLiteral]
+from typing_extensions import Literal
+def foo(x: Literal[1, 2, 3]) -> None: pass
+
+a: Literal[1]
+b: Literal[2, 3]
+c: Literal[4, 5]
+d: int
+
+foo(a)
+foo(b)
+foo(c)  # E: Argument 1 to "foo" has incompatible type "Union[Literal[4], Literal[5]]"; expected "Union[Literal[1], Literal[2], Literal[3]]"
+foo(d)  # E: Argument 1 to "foo" has incompatible type "int"; expected "Union[Literal[1], Literal[2], Literal[3]]"
+[out]
+
+[case testLiteralCallingFunctionWithStandardBase]
+from typing_extensions import Literal
+def foo(x: int) -> None: pass
+
+a: Literal[1]
+b: Literal[1, -4]
+c: Literal[4, 'foo']
+
+foo(a)
+foo(b)
+foo(c)  # E: Argument 1 to "foo" has incompatible type "Union[Literal[4], Literal['foo']]"; expected "int"
+[out]
+
+[case testLiteralCallingOverloadedFunction]
+from typing import overload, Generic, TypeVar, Any
+from typing_extensions import Literal
+
+T = TypeVar('T')
+class IOLike(Generic[T]): pass
+
+@overload
+def foo(x: Literal[1]) -> IOLike[int]: ...
+@overload
+def foo(x: Literal[2]) -> IOLike[str]: ...
+@overload
+def foo(x: int) -> IOLike[Any]: ...
+def foo(x: int) -> IOLike[Any]:
+    if x == 1:
+        return IOLike[int]()
+    elif x == 2:
+        return IOLike[str]()
+    else:
+        return IOLike()
+
+a: Literal[1]
+b: Literal[2]
+c: int
+d: Literal[3]
+
+reveal_type(foo(a))  # E: Revealed type is '__main__.IOLike[builtins.int]'
+reveal_type(foo(b))  # E: Revealed type is '__main__.IOLike[builtins.str]'
+reveal_type(foo(c))  # E: Revealed type is '__main__.IOLike[Any]'
+foo(d)
+[builtins fixtures/ops.pyi]
```

Comment:
Add tests for `Any`, `NoReturn`, tests involving invariance/covariance (`List[Literal[1, 2]]` vs `Sequence[Literal[1, 2]]`), and `None` with `strict_optional = False`.